vowpalwabbit/array_parameters.h

```diff
@@ -0,0 +1,172 @@
+
+class weights_iterator_iterator
+{
+private:
+	weight* _cur;
+public:
+	weights_iterator_iterator(weight* cur)
+		: _cur(cur)
+	{ }
+	
+	weight& operator*() { return *_cur; }
+
+	weights_iterator_iterator& operator++()
+	{
+		_cur++;
+		return *this;
+	}
+
+	template<typename T>
+	weights_iterator_iterator operator+(T index) { return weights_iterator_iterator(_cur + index); }
+
+	template<typename T>
+	weights_iterator_iterator& operator+=(T index)
+	{
+		_cur += index;
+		return *this;
+	}
+
+	weights_iterator_iterator& operator=(const weights_iterator_iterator& other)
+	{
+		_cur = other._cur;
+		return *this;
+	}
+
+	bool operator==(const weights_iterator_iterator& rhs) { return _cur == rhs._cur; }
+	bool operator!=(const weights_iterator_iterator& rhs) { return _cur != rhs._cur; }
+
+};
+class weights_iterator
+{
+private:
+	weight* _current;
+	uint32_t _stride;
+
+public:
+	typedef weights_iterator_iterator w_iter;
+
+	weights_iterator(weight* current, uint32_t stride)
+		: _current(current), _stride(stride)
+	{ }
+
+	weight& operator*() { return *_current; }
+
+	weights_iterator& operator++()
+	{
+		_current += _stride;
+		return *this;
+	}
+
+	template<typename T>
+	weights_iterator operator+(T index) { return weights_iterator(_current + (index*_stride), _stride); }
+
+	template<typename T>
+	weights_iterator& operator+=(T index)
+	{
+		_current += (index*_stride);
+		return *this;
+	}
+
+	weights_iterator& operator=(const weights_iterator& other)
+	{
+		_current = other._current;
+		_stride = other._stride;
+		return *this;
+	}
+
+	bool operator==(const weights_iterator& rhs) { return _current == rhs._current; }
+	bool operator!=(const weights_iterator& rhs) { return _current != rhs._current; }
+
+	//to iterate within a bucket
+	w_iter begin() { return w_iter(_current); }
+	w_iter end(size_t offset) { return w_iter(_current + offset); }
+
+};
+
+class weight_vector //different name? (previously array_parameters)
+{
+private:
+	weight* _begin;
+	uint64_t _weight_mask;  // (stride*(1 << num_bits) -1)
+	uint32_t _stride_shift;
+	bool _seeded;
+
+public:
+	typedef weights_iterator iterator;
+
+	weight_vector()
+		: _begin(nullptr), _weight_mask((uint64_t)LONG_MAX), _stride_shift(0),_seeded(false)
+	{ }
+
+	weight_vector(size_t length, uint32_t stride_shift=0)
+		: _begin(calloc_mergable_or_throw<weight>(length << stride_shift)),
+		_weight_mask((length << stride_shift) - 1),	
+		_stride_shift(stride_shift),
+		_seeded(false)
+	{ }
+
+	inline weight* first() { return _begin; } //TODO: Temporary fix for lines like (&w - all.reg.weight_vector). Needs to change for sparse.
+	iterator begin() { return iterator(_begin, 1); }
+	iterator end() { return iterator(_begin + _weight_mask + 1, 1); }
+
+	//iterator with stride and offset
+	iterator begin(size_t offset) { return iterator(_begin + offset, (1<<_stride_shift)); }
+	iterator end(size_t offset) { return iterator(_begin + _weight_mask + 1 + offset, (1 << _stride_shift)); }
+	
```

Comment:
I have a number of comments regarding the iterator interfaces.  First of all, you should consider adding a const_iterator const versions of these functions, so programmers can express their desire to only read the data.   Also, consider providing a range interface for your `begin` and `end` (especially with offsets), so it's possible to write something like `for (auto val: vw.read(offset)) { }`.  They say that one shouldn't repeat themselves in code, and having to call `begin(offset)` and `end(offset)` certainly qualifies as that.

Finally, one more thing: Your rewrite of vw using your weight_vector class has brought up an interesting observation:  There are a lot of instances of loops which only iterate over a specific offset.  Now that you have this beautiful abstraction, you may want to consider the cache behaviour of your code.  If you're always iterating a specific offset in a bucket, it would be much more efficient to bucket the offsets together, i.e. instead of

w11 w12 w13 w21 w22 w23 ...

you could do

w11 w21 ... w12 w22 ... w13 w23...

As far as your clients are concerned, nothing changes because you are still presenting the same interface, but the computation could be a lot faster.

From my cursory review of VW, it seems that the layout may need to be algorithm dependent, but I still thought it's worth mentioning.